accepted/0000-retail-pxe-database.md

```diff
@@ -0,0 +1,257 @@
+
+
+To be able to recreate PXE entries, it is needed to store some additional details about terminals, make them available for the branch server minion and have a state or other mechanism which will recreate them. Idea is that after replacing branch server for any reason, user is able to apply a state or schedule an action and recreate the PXE entries for all connected terminals.
+
+By PXE entries are meant both PXE configuration files commonly named `01-mac_address_dash_separated` and GRUB configuration for UEFI commonly named `01:mac_address_colon_separated.grub2.cfg`.
+
+Later in the text I introduced term `pxe server`. This is usually branch server proxy, however design should be flexible enough to allow any salt client to become pxe server, particularly in branchless scenarios with e.g. peripheral hub servers instead of branch server proxy.
+
+# Detailed design
+[design]: #detailed-design
+
+
+## Workflow description
+
+On each boot (first boot and all next boots) terminal sends an event `suse/manager/pxe_update` together with following data:
+
+```
+salt_device: $dev_by_id_path
+root: $dev_by_id_path
+boot_image: $boot_image_name
+terminal_kernel_parameters: $hwtype_based_parameters
+```
+
+These data are then, together with data about branch (provided on kernel command line), used to create PXE and GRUB entries like:
+
+```
+LABEL netboot
+        kernel POS_Image_JeOS7-7.0.0/POS_Image_JeOS7.x86_64-7.0.0-5.3.18-24.37-default.kernel
+        append initrd=POS_Image_JeOS7-7.0.0/POS_Image_JeOS7.x86_64-7.0.0.initrd.xz  panic=60 ramdisk_size=710000 ramdisk_blocksize=4096 vga=0x317 splash=silent  USE_FQDN_MINION_ID=1 MINION_ID_PREFIX=branch1 root=/dev/disk/by-path/pci-0000:04:00.0-part3 salt_device=/dev/disk/by-path/pci-0000:04:00.0-part3
+```
+
+and
+
+```
+menuentry netboot {
+        linuxefi ${prefix}/POS_Image_JeOS7-7.0.0/POS_Image_JeOS7.x86_64-7.0.0-5.3.18-24.37-default.kernel panic=60 ramdisk_size=710000 ramdisk_blocksize=4096 vga=0x317 splash=silent  USE_FQDN_MINION_ID=1 MINION_ID_PREFIX=branch1 root=/dev/disk/by-path/pci-0000:04:00.0-part3 salt_device=/dev/disk/by-path/pci-0000:04:00.0-part3
+        initrdefi ${prefix}/POS_Image_JeOS7-7.0.0/POS_Image_JeOS7.x86_64-7.0.0.initrd.xz
+}
+```
+
+The data send in the `pxe_update` event are not stored anywhere on the Uyuni/SUSE Manager and upon receiving the event, salt reactor applies `pxe/terminal_entry` state on the branch server which will create the PXE entry for given terminal.
+
+
+To be able to recreate PXE entries at later time, it is clear that the data send in the `pxe_update` event should be stored somewhere on the Uyuni/SUSE Manager and be accessible to branch servers.
+
+
+## Storage of the data
```

Comment:
Alternative idea:
 - we provide a mechanism to save arbitrary pillar data (eg. an arbitrary JSON blob) into the database, per minion. This comes with Java methods to read and write such data
 - we provide a way (likely implemented with some new API) to consume such DB-backed pillars (via a new [external pillar](https://docs.saltstack.com/en/latest/topics/development/modules/external_pillars.html) module)
 - code is added to the Java app to react to `suse/manager/pxe_update`. This code will save data in the event as pillar data into the database
 - a new state is written that reads such pillar data from the external pillar (thus the DB) and writes PXE/GRUB entries. This can be triggered on highstate, by reactor on `suse/manager/pxe_update` or manually via normal means

How does that sound?

Reply:
This sounds good 👍🏾

Reply:
I'm preparing JSON blob storage as a separate RFC draft and there are some pain points wrt pxe entries.

Pxe entries pillars are meant to be available to branch proxy.
What exactly did you mean by `have arbitrary pillar data into the database, per minion`?
* at most one entry per minion? In this case we would need to do JSON preprocessing to concatenate/remove entries as needed when modifying it.
* there can be multiple blobs per minion? Removing/modifying entries would be easier, merging would be done in external pillar/salt.

Next these pxe entries are closely linked with the terminal minions, however they are not meant for them. So if terminal minion is removed, its pxe entry should be removed as well (not a hard requirement, but under some circumstances can lead to error during deployment).
To support this scenario we can either:
- have an owner of the JSON blob for whom the pillars are meant and then a list of systems on which it depends. I don't know if there is any other use case for this architecture and this would require to have multiple blobs per minion
- adapt system entry removal procedure to check if it is a retail minion (e.g. member of TERMINAL groups), find its associated proxy and modify branch proxy pxe entries JSON blob.

Reply:
> I'm preparing JSON blob storage as a separate RFC draft and there are some pain points wrt pxe entries.
> 
> Pxe entries pillars are meant to be available to branch proxy.
> What exactly did you mean by `have arbitrary pillar data into the database, per minion`?

Just to make sure we are on the same page: in Salt, [a pillar](https://docs.saltstack.com/en/latest/topics/tutorials/pillar.html#pillar-walkthrough) is a "JSON-like" arbitrary data structure _relative to a certain minion ID_. "JSON-like" here means it can basically be strings, integers, booleans, none plus lists and dictionaries (however nested). Effectively, any pillar can be serialized into JSON, and it's typically represented in YAML (which is a superset of JSON).

An "external pillar" is a source of pillar data that comes from a third party. Could be a database, LDAP, Amazon S3, or anything really. All one needs to do in order to implement such an "external pillar" is to define [a function that returns the data given a minion ID](https://docs.saltstack.com/en/latest/topics/development/modules/external_pillars.html#ext-pillar) (and a few other secondary implementation details).

So the idea above is really to provide an Uyuni-backed external pillar, where pillar data is stored inside of the database (for each minion).

>     * at most one entry per minion? In this case we would need to do JSON preprocessing to concatenate/remove entries as needed when modifying it.
>     * there can be multiple blobs per minion? Removing/modifying entries would be easier, merging would be done in external pillar/salt.

The easiest implementation I can think of is to have the JSON blob in the database mapped to Java objects, so that any manipulation happens in Java code where it's easiest to write/debug/test. Basically, Java would load the blob as-is, alter it, and then re-save it to the DB as a whole. In that case, the first option seems straightforward.

Of course that could bring to performance problems, in case we need to manipulate pillars of very many minions, or in case blobs become very big. In that case, maybe we could do the manipulation [directly in SQL](https://www.postgresql.org/docs/11/functions-json.html), such as [removing a key](https://stackoverflow.com/questions/23490965/postgresql-remove-attribute-from-json-column) or setting one, even deep in the structure. This could be done as an incremental improvement to the easiest implementation above, even at a later point in time.

So I do not see big advantages with a 1 minion to N blobs approach - but maybe I'm missing something.

> Next these pxe entries are closely linked with the terminal minions, however they are not meant for them. So if terminal minion is removed, its pxe entry should be removed as well (not a hard requirement, but under some circumstances can lead to error during deployment).

How is the cleanup performed in the current implementation?

> To support this scenario we can either:
> 
>     * have an owner of the JSON blob for whom the pillars are meant and then a list of systems on which it depends. I don't know if there is any other use case for this architecture and this would require to have multiple blobs per minion

I cannot presently think of other possible cases, so I would be more inclined to call this a Retail-specific mechanism until we find a second user. Only then I would market it as general.

>     * adapt system entry removal procedure to check if it is a retail minion (e.g. member of TERMINAL groups), find its associated proxy and modify branch proxy pxe entries JSON blob.

Idea: when a minion is deleted, its minion key must be deleted, and Salt sends out an event in that case. Would it be an option to react to such event to do the cleanup, just as we react to `suse/manager/pxe_update` to start this process? (as event reaction is idiomatic in Salt, I guess this would be the idiomatic way to resolve the problem in the Salt world)

Yet another alternative: how about a Taskomatic cleanup job that checks all the Retail Proxies and cleans up any pillars belonging to clients that are not there any more? (As garbage collection in the database via Taskomatic is idiomatic in Spacewalk, I guess this would be the idiomatic way to resolve the problem in the Spacewalk world)

Reply:
> Just to make sure we are on the same page: in Salt, [a pillar](https://docs.saltstack.com/en/latest/topics/tutorials/pillar.html#pillar-walkthrough) is a "JSON-like" arbitrary data structure _relative to a certain minion ID_. "JSON-like" here means it can basically be strings, integers, booleans, none plus lists and dictionaries (however nested). Effectively, any pillar can be serialized into JSON, and it's typically represented in YAML (which is a superset of JSON).
> 
> An "external pillar" is a source of pillar data that comes from a third party. Could be a database, LDAP, Amazon S3, or anything really. All one needs to do in order to implement such an "external pillar" is to define [a function that returns the data given a minion ID](https://docs.saltstack.com/en/latest/topics/development/modules/external_pillars.html#ext-pillar) (and a few other secondary implementation details).

That is how I understood it.

> > ```
> > * at most one entry per minion? In this case we would need to do JSON preprocessing to concatenate/remove entries as needed when modifying it.
> > * there can be multiple blobs per minion? Removing/modifying entries would be easier, merging would be done in external pillar/salt.
> > ```
> 
> The easiest implementation I can think of is to have the JSON blob in the database mapped to Java objects, so that any manipulation happens in Java code where it's easiest to write/debug/test. Basically, Java would load the blob as-is, alter it, and then re-save it to the DB as a whole. In that case, the first option seems straightforward.
> 
> Of course that could bring to performance problems, in case we need to manipulate pillars of very many minions, or in case blobs become very big. In that case, maybe we could do the manipulation [directly in SQL](https://www.postgresql.org/docs/11/functions-json.html), such as [removing a key](https://stackoverflow.com/questions/23490965/postgresql-remove-attribute-from-json-column) or setting one, even deep in the structure. This could be done as an incremental improvement to the easiest implementation above, even at a later point in time.

👍🏾 
 
> > Next these pxe entries are closely linked with the terminal minions, however they are not meant for them. So if terminal minion is removed, its pxe entry should be removed as well (not a hard requirement, but under some circumstances can lead to error during deployment).
> 
> How is the cleanup performed in the current implementation?

It is not done curently, something I would like to fix with this development as well.

> Idea: when a minion is deleted, its minion key must be deleted, and Salt sends out an event in that case. Would it be an option to react to such event to do the cleanup, just as we react to `suse/manager/pxe_update` to start this process? (as event reaction is idiomatic in Salt, I guess this would be the idiomatic way to resolve the problem in the Salt world)
> 
> Yet another alternative: how about a Taskomatic cleanup job that checks all the Retail Proxies and cleans up any pillars belonging to clients that are not there any more? (As garbage collection in the database via Taskomatic is idiomatic in Spacewalk, I guess this would be the idiomatic way to resolve the problem in the Spacewalk world)

I wasn't aware salt sends out an event on minion key deletion. That sounds interesting and good way to trigger this deletion on branch proxy when minion is deleted.